moto/ec2/models.py

```diff
@@ -1155,14 +1155,15 @@ def add_instances(self, image_id, count, user_data, security_group_names, **kwar
                         "DeleteOnTermination", False
                     )
                     kms_key_id = block_device["Ebs"].get("KmsKeyId")
-                    new_instance.add_block_device(
-                        volume_size,
-                        device_name,
-                        snapshot_id,
-                        encrypted,
-                        delete_on_termination,
-                        kms_key_id,
-                    )
+                    if block_device["NoDevice"] != "":
```

Comment:
What is the AWS behaviour when the user specifies a value other than an empty string, e.g. `NoDevice="yes"`? The docs are pretty clear that it should be `None` or `""`, but it doesn't say what happens for other values.

Reply:
I was wondering the same thing.  i'll try it out and I was thinking of adding the appropriate exception to my moto PR if boto throws in that case.

Reply:
If I pass `'NotAnEmptyString'` I get the following stacktrace:
```
  File "/home/centos/firesim_oss_220203/deploy/awstools/awstools.py", line 261, in launch_instances
    instance = ec2.create_instances(**instance_args)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/boto3/resources/factory.py", line 520, in do_action
    response = action(self, *args, **kwargs)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(*args, **params)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/botocore/client.py", line 391, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/botocore/client.py", line 719, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (InvalidRequest) when calling the RunInstances operation: The request received was invalid.
```

And passing `None` I get a different stacktrace:
```
  File "/home/centos/firesim_oss_220203/deploy/awstools/awstools.py", line 261, in launch_instances
    instance = ec2.create_instances(**instance_args)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/boto3/resources/factory.py", line 520, in do_action
    response = action(self, *args, **kwargs)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(*args, **params)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/botocore/client.py", line 391, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/botocore/client.py", line 692, in _make_api_call
    api_params, operation_model, context=request_context)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/botocore/client.py", line 740, in _convert_to_request_dict
    api_params, operation_model)
  File "/home/centos/miniforge3/envs/firesim_oss_220203/lib/python3.7/site-packages/botocore/validate.py", line 360, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter BlockDeviceMappings[1].NoDevice, value: None, type: <class 'NoneType'>, valid types: <class 'str'>
```

I'm not sure where you saw that `None` could be passed.  The docs I was looking at only said that it could be the empty string but I think I was looking that the AWS api docs, not boto.

I'm happy to add validation to my PR that throws similar/same exceptions for those cases.

Reply:
Thanks.  I made that change and it looks like everything is passing now.